crates/topos-tce/src/app_context.rs

```diff
@@ -209,6 +209,41 @@ impl AppContext {
 
                 _ = sender.send(result);
             }
+
+            ApiEvent::GetLastPendingCertificates {
+                mut subnet_ids,
+                sender,
+            } => {
+                let mut last_pending_certificates: HashMap<String, Option<Certificate>> =
+                    subnet_ids
+                        .iter()
+                        .map(|subnet_id| (subnet_id.to_string(), None))
+                        .collect();
+
+                if let Ok(pending_certificates) =
+                    self.pending_storage.get_pending_certificates().await
+                {
+                    // Iterate through pending certificates and determine last one for every subnet
+                    // Last certificate in the subnet should be one with the highest index
+                    for (pending_certificate_id, cert) in pending_certificates.into_iter().rev() {
```

Comment:
we probably want to rely on the field `position` of the cert rather than on the order of the column, since the column is inserting for a FIFO access I think, so normally nothing prevents one peer to insert an older cert after a more recent one (in terms of position). We could have it as a separate ticket if it has indeed to be done :ok_hand:

Reply:
There is no `position` on a `pending_certificates` and we can't enforce ordering before the broadcast happens. There is no clear solution to do what you suggest. The whole `get last pending` features is regarding a TCE node state, not a sequencer. It's up to the sequencer to decide whether it can trust this TCE or not.

Reply:
@hadjiszs Position is defined when you deliver the certificate - as implemented now, pending certificates do not have position. We can discuss this topic when we further refactor pending certificates handling

Reply:
Right, I mentioned about the field `position` but should refer to the field `prev_id` instead. The ordering of a source stream is already defined prior to the broadcast, thanks to the `prev_id`. We are not talking about the ordering on the point of view of a target subnet here.

Given the source subnet A

```
Cert 0x01 {
  source: subnet_a,
  prev_id: 0x00
}

Cert 0x02 {
  source: subnet_a,
  prev_id: 0x01
}
```

Here we could have 0x01 after 0x02 in our pending column, and we'll send `0x01` as the last pending for `subnet_a`, while we should send `0x02` instead. And we have the data to know that the latest one is 0x02 thanks to the `prev_id`, we don't need to proceed to the broadcast to know that.

The problem happens if there is holes in our pending column, e.g. we have `0x00` and `0x02` but not `0x01`, then we can't know the ordering between `0x00` and `0x02`, so in that case it sounds legit to take the last one in the column.

Also, it's not about trust issues here, just about the data itself that is returned and tagged as "last pending", which is not in the point of view of the TCE node itself if we rely only on the order of the column of the pending.

Reply:
@hadjiszs this is a known issue and the main problem is the one you're describing. I think that we'll need to consume and requeue the certificates that fall into this hole.